### MOMA dataset!

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
collection = pd.read_csv('https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artists.csv')

In [3]:
collection.columns

Index(['ConstituentID', 'DisplayName', 'ArtistBio', 'Nationality', 'Gender',
       'BeginDate', 'EndDate', 'Wiki QID', 'ULAN'],
      dtype='object')

In [4]:
# Select Columns.
collection = collection[['ConstituentID', 'DisplayName', 'ArtistBio', 'Nationality', 'Gender',
       'BeginDate', 'EndDate', 'Wiki QID', 'ULAN']]





# Drop missing data.
collection = collection.dropna()

In [5]:
collection.head()

ConstituentID          DisplayName  \
3               4      Charles Arnoldi   
7               9        David Aronson   
8              10        Irene Aronson   
9              11      Jean (Hans) Arp   
15             19  Richard Artschwager   

                                   ArtistBio Nationality  Gender  BeginDate  \
3                        American, born 1946    American    Male       1946   
7              American, born Lithuania 1923    American    Male       1923   
8                American, born Germany 1918    American  Female       1918   
9   French, born Germany (Alsace). 1886–1966      French    Male       1886   
15                       American, 1923–2013    American    Male       1923   

    EndDate   Wiki QID         ULAN  
3         0   Q1063584  500027998.0  
7         0   Q5230870  500003363.0  
8         0  Q19748568  500042413.0  
9      1966    Q153739  500031000.0  
15     2013    Q568262  500114981.0

In [6]:
# Get data types.
collection.dtypes

ConstituentID      int64
DisplayName       object
ArtistBio         object
Nationality       object
Gender            object
BeginDate          int64
EndDate            int64
Wiki QID          object
ULAN             float64
dtype: object

In [7]:


# Final column drops and NA drop.
X = collection.drop(['Wiki QID', 'DisplayName', 'ArtistBio', 'Nationality', 'Gender'], 1)

# Create dummies separately.
gender = pd.get_dummies(collection.Gender)


# Concat with other variables, but artists slows this wayyyyy down so we'll keep it out for now
X = pd.get_dummies(X, sparse=True)
X = pd.concat([X, gender], axis=1)

Y = collection.Nationality

In [8]:
from sklearn.model_selection import train_test_split


X_train, X_test, Y_train, Y_test = train_test_split(
    X,
    Y,
    test_size=0.9,
    random_state=42)

In [9]:
# Alright! We've done our prep, let's build the model.
# Neural networks are hugely computationally intensive.
# This may take several minutes to run.

# Import the model.
from sklearn.neural_network import MLPClassifier

# Establish and fit the model, with a single, 1000 perceptron layer.
mlp = MLPClassifier(hidden_layer_sizes=(2000,))
mlp.fit(X, Y)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(2000,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [11]:
Y.value_counts()/len(Y)

American         0.421016
French           0.097425
British          0.096729
German           0.093946
Italian          0.043145
Japanese         0.026444
Swiss            0.023660
Dutch            0.018441
Austrian         0.017745
Russian          0.013918
Canadian         0.012874
Mexican          0.012526
Belgian          0.011830
Spanish          0.011482
Brazilian        0.009743
Polish           0.008699
Argentine        0.006959
Swedish          0.005567
Czech            0.005219
Danish           0.005219
Israeli          0.004523
Australian       0.004175
Finnish          0.003827
Venezuelan       0.002784
South African    0.002784
Hungarian        0.002784
Cuban            0.002436
Chilean          0.002436
Chinese          0.002436
Colombian        0.002088
                   ...   
Guatemalan       0.000696
Turkish          0.000696
Scottish         0.000696
Iranian          0.000696
Croatian         0.000696
Slovak           0.000696
Korean           0.000696
Slovenian   

In [12]:
Y.value_counts()/len(Y)

American         0.421016
French           0.097425
British          0.096729
German           0.093946
Italian          0.043145
Japanese         0.026444
Swiss            0.023660
Dutch            0.018441
Austrian         0.017745
Russian          0.013918
Canadian         0.012874
Mexican          0.012526
Belgian          0.011830
Spanish          0.011482
Brazilian        0.009743
Polish           0.008699
Argentine        0.006959
Swedish          0.005567
Czech            0.005219
Danish           0.005219
Israeli          0.004523
Australian       0.004175
Finnish          0.003827
Venezuelan       0.002784
South African    0.002784
Hungarian        0.002784
Cuban            0.002436
Chilean          0.002436
Chinese          0.002436
Colombian        0.002088
                   ...   
Guatemalan       0.000696
Turkish          0.000696
Scottish         0.000696
Iranian          0.000696
Croatian         0.000696
Slovak           0.000696
Korean           0.000696
Slovenian   

In [13]:
from sklearn.model_selection import cross_val_score
cross_val_score(mlp, X, Y, cv=5)

/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


array([0.39477977, 0.41367521, 0.0122807 , 0.10071942, 0.44      ])

In [10]:
mlp.score(X, Y)

0.012526096033402923

### Ok so we got an MLP score of .0125, now let's comparing with Random Forest 

In [16]:
from sklearn.model_selection import train_test_split


In [19]:
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn import ensemble
# Final column drops and NA drop.
X = collection.drop(['Wiki QID', 'DisplayName', 'ArtistBio', 'Nationality', 'Gender'], 1)

# Create dummies separately.
gender = pd.get_dummies(collection.Gender)


# Concat with other variables, but artists slows this wayyyyy down so we'll keep it out for now
X = pd.get_dummies(X, sparse=True)
X = pd.concat([X, gender], axis=1)

Y = collection.Nationality

randomforest = RandomForestRegressor()
randomforest.fit(X_train, Y_train)
randomforest.score(X_test, Y_test)

/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


ValueError: could not convert string to float: 'Slovenian'